### PR TITLE
Adding vpn connection resource

### PR DESCRIPTION
**Pull request by Rohit1509**
### Description

Adding vpn connection resource

### Issues Resolved

List any existing issues this PR resolves, or any Discourse or StackOverflow discussion that's relevant

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [ ] New functionality includes integration tests/controls
- [ ] New Terraform resources
- [ ] Documentation provided or updated for resources 
- [ ] All Integration Tests pass
- [ ] All Unit Tests pass
- [ ] `rake lint` passes
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by soumyo13** [comment]
@Rohit1509 Please add the integration test and the documentation.

File: libraries/aws_vpn_connection.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  def exists?
+    !failed_resource? && !!@response || !@response.empty?
+  end
+
+  def to_s
+    "VPN connection ID: #{@opts[:vpn_connection_id]}"
+  end
+
+  def resource_id
+    @opts ? @opts[:vpn_connection_id] : ''
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:36]: We should not pass an empty string. If we pass an empty string then the resource_id will be empty. 

We can @response[: vpn_connection_id]

**Comment by soumyo13** [comment]
Here only singular resource is added. Can you please add the plural resource also.

**Comment by Rohit1509** [comment]
> Here only singular resource is added. Can you please add the plural resource also.

Plural resource already exists

File: libraries/aws_vpn_connection.rb (hunk 1 above)
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:36]: @response ? @response.vpn_connections.first.vpn_connection_id : @display_name

File: libraries/aws_vpn_connection.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+   end
+
+   describe aws_vpn_connection('vc-014aef8a0689b8f43') do
+     it { should exist }
+   end
+  "
+
+  def initialize(opts = {})
+    opts = { vpn_connection_id: opts } if opts.is_a?(String)
+    super
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:21]:
```suggestion
    super(opts)
```

File: libraries/aws_vpn_connection.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    validate_parameters(required: [:vpn_connection_id])
+    validate_identifier
+    fetch
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:24]:
```suggestion
    validate_identifier
    @display_name = @opts[:vpn_connection_id]
    fetch
```

File: libraries/aws_vpn_connection.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def exists?
+    !failed_resource? && !!@response || !@response.empty?
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:28]:
```suggestion
    !@response.blank?
```

File: libraries/aws_vpn_connection.rb (hunk 1 above)
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:36]:
```suggestion
    @response ? @response.vpn_connections.first.vpn_connection_id : @display_name
```

File: libraries/aws_vpn_connection.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def to_s
+    "VPN connection ID: #{@opts[:vpn_connection_id]}"
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:32]:
```suggestion
    "VPN Connection ID: #{@display_name}"
```

File: libraries/aws_vpn_connection.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def resource_id
+    @opts ? @opts[:vpn_connection_id] : ''
+  end
+
+  private
+
+  def fetch
+    params = { vpn_connection_ids: [@opts[:vpn_connection_id]] }
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:42]:
```suggestion
    params = { vpn_connection_ids: [@display_name] }
```

File: libraries/aws_vpn_connection.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def query_with(params)
+    catch_aws_errors do
+      @aws.compute_client.describe_vpn_connections(params)
+    end
+  end
+
+  def validate_identifier
+    raise ArgumentError, 'parameter `vpn_connection_id` cannot be blank' if @opts[:vpn_connection_id].nil? || @opts[:vpn_connection_id].empty?
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:56]:
```suggestion
    raise ArgumentError, 'parameter `vpn_connection_id` cannot be blank' if @opts[:vpn_connection_id].blank?
```

File: libraries/aws_vpn_connection.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  desc 'Verifies settings for a single AWS VPN Connection.'
+
+  example "
+   describe aws_vpn_connection(vpn_connection_id: 'vc-014aef8a0689b8f43') do
+     it { should exist }
+   end
+
+   describe aws_vpn_connection('vc-014aef8a0689b8f43') do
+     it { should exist }
+   end
```
Comment (soumyo13) [inline libraries/aws_vpn_connection.rb:16]:
```suggestion
   describe aws_vpn_connection(vpn_connection_id: 'vpn-1234567890') do
     it { should exist }
   end

   describe aws_vpn_connection('vpn-1234567890') do
     it { should exist }
   end
```

File: test/unit/resources/aws_vpn_connection_test.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    assert @connection.exists?
+  end
+
+  def test_id
+    assert_equal(@connection.vpn_connection_id, @vpn_connection.vpn_connection_id)
+  end
+
+  def test_resource_id
+    refute_nil(@connection.resource_id)
+    assert_equal(@connection.resource_id, @connection.vpn_connection_id)
```
Comment (soumyo13) [inline test/unit/resources/aws_vpn_connection_test.rb:41]:
```suggestion
    assert_equal(@connection.resource_id, @vpn_connection.vpn_connection_id)
```

**Comment by Rohit1509** [comment]
@soumyo13 all the requested changes have been made. Thanks!

**Comment by Rohit1509** [comment]
Hi @soumyo13 , can you please look into why the PR validation is failing ? It looks to be out of scope of this PR. Thanks!

**Review by sathish-progress — APPROVED** [review]
LGTM

**Review by soumyo13 — APPROVED** [review]

**Review by sathish-progress — APPROVED** [review]

**Pull request merged**